ilp-core/src/main/java/org/interledger/core/InterledgerAddressParser.java

```diff
@@ -0,0 +1,133 @@
+package org.interledger.core;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * An {@link org.interledger.core.InterledgerAddress}es parser.
+ */
+final class InterledgerAddressParser {
+
+  static enum Error {
+    ADDRESS_OVERFLOW("Address is too long"),
+    INVALID_SEGMENT("The '%s' segment has an invalid format"),
+    INVALID_SCHEME_PREFIX("The '%s' scheme prefix has an invalid format"),
+    MISSING_SCHEME_PREFIX("Address does not start with a scheme prefix"),
+    SEGMENTS_UNDERFLOW("Destination address has too few segments");
+
+    private String messageFormat;
+
+    private Error(final String messageFormat) {
+      this.messageFormat = messageFormat;
+    }
+
+    public String getMessageFormat() {
+      return messageFormat;
+    }
+  }
+
+  private static final int MIN_ADDRESS_LENGTH = 1;
+  private static final int MAX_ADDRESS_LENGTH = 1023;
+  private static final int DESTINATION_ADDRESS_MIN_SEGMENTS = 2;
+
+  private static final String SCHEME_REGEX = "(g|private|example|peer|self|test[1-3]?)";
+  private static final String SEGMENT_REGEX = "[a-zA-Z0-9_~-]+";
+  private static final String SEPARATOR_CHARACTER = ".";
+  private static final String SEPARATOR_REGEX = "[" + SEPARATOR_CHARACTER + "]";
+  private static final String SCHEME_PREFIX_REGEX = SCHEME_REGEX + SEPARATOR_REGEX;
+  private static final String SEGMENT_PREFIX_REGEX = SEGMENT_REGEX + SEPARATOR_REGEX;
+  private static final String ADDRESS_LENGTH_BOUNDARIES_REGEX = "(?=^.{" + MIN_ADDRESS_LENGTH + ","
+      + MAX_ADDRESS_LENGTH + "}$)";
+
+  private static final String ADDRESS_PREFIX_REGEX = ADDRESS_LENGTH_BOUNDARIES_REGEX
+      + "^" + SCHEME_PREFIX_REGEX + "(" + SEGMENT_PREFIX_REGEX + ")*$";
+  private static final Pattern ADDRESS_PREFIX_PATTERN = Pattern.compile(ADDRESS_PREFIX_REGEX);
+
+  private static final String DESTINATION_ADDRESS_REGEX = ADDRESS_LENGTH_BOUNDARIES_REGEX
+      + "^" + SCHEME_PREFIX_REGEX + "(" + SEGMENT_PREFIX_REGEX + ")+"
+      + SEGMENT_REGEX + "$";
+  private static final Pattern DESTINATION_ADDRESS_PATTERN = Pattern.compile(
+      DESTINATION_ADDRESS_REGEX
+  );
+
+  private static final String SCHEME_PREFIX_ONLY_REGEX = "^" + SCHEME_PREFIX_REGEX + "$";
+  private static final Pattern SCHEME_PREFIX_ONLY_PATTERN = Pattern.compile(
+      SCHEME_PREFIX_ONLY_REGEX
+  );
+
+  /**
+   * Validates an ILP address.
+   * 
+   * @param addressString The ILP address to validate
+   */
+  void validate(final String addressString) {
```

Comment:
Do you think that the method signature - as well as its Javadoc - should state `throws IllegalArgumentException`?

I think so, even if mentionning unchecked exceptions in method signatures is bad practice (because most of time, they are not checked/catched by caller). This way, a dev doesn't have to dig into the method's logic to finally find that validation gets rejected through a `RuntimeException`.

Reply:
Sounds fine to me.

Reply:
This point is now adressed.